x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/SecurityContext.java

```diff
@@ -150,6 +151,16 @@ public void executeAsUser(User user, Consumer<StoredContext> consumer, Version v
         }
     }
 
+    /**
+     * Compared to {@code ThreadContext#stashContext} this only clears the specific transient headers that are used to convey the
+     * authorization outcome, leaving the other headers in place, just like {@code ThreadContext#newStoredContext(true)}. When
+     * the returned {@code StoredContext} is closed, the original authorization transient headers are restored (if any did
+     * not originally exist, it will also not exist after the restore).
+     */
+    public StoredContext stashAuthorizationContext() {
+        return threadContext.stashTransientContext(AuthorizationServiceField.ALL_AUTHORIZATION_KEYS);
+    }
```

Comment:
I am not sure whether it is necessary to package this into `SecurityContext` class. I'd prefer to have it as a method of `AuthorizationService`, since it is the only consumer. Therefore it will be close to the context if it is defined in `AuthorizationService`. Based on the method name, it is also unlikley to be used anywhere else in future.

Wrapping it here leads to many boilerplate code like `securityContext.getThreadContext()...`, which does not help reading and maybe obsecure the actual importance being only "transient header stashing" for future readers.

Reply:
I still maintain that it sits better in the `SecurityContext` even if the `AuthorizationService` is the single caller, but the topic is not significant enough to argue about, so I went with your suggestion.